core/src/main/java/com/datastax/oss/driver/api/core/metrics/Metrics.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.api.core.metrics;
+
+import com.codahale.metrics.Metric;
+import com.codahale.metrics.MetricRegistry;
+import com.datastax.oss.driver.api.core.metadata.Node;
+import com.datastax.oss.driver.api.core.type.reflect.GenericType;
+
+/**
+ * A wrapper around a {@link MetricRegistry} to expose the driver's metrics.
+ *
+ * <p>This type exists mainly to avoid a hard dependency to Dropwizard Metrics (that is, the JAR can
+ * be completely removed from the classpath if metrics are disabled). It also provides convenience
+ * methods to access individual metrics programatically.
+ */
+public interface Metrics {
+
+  /**
+   * Returns the underlying Dropwizard registry.
+   *
+   * <p>Typically, this can be used to configure a reporter.
+   *
+   * @see <a href="http://metrics.dropwizard.io/4.0.0/manual/core.html#reporters">Reporters
+   *     (Dropwizard Metrics manual)</a>
+   */
+  MetricRegistry getRegistry();
```

Comment:
This references Dropwizard classes. This means that if someone were to integrate another metrics framework, they would have to make `session.getMetrics` return `Optional.empty()`,  and find another way to expose that framework's equivalent of the registry.
I don't want to go further than that, because any other solution would have to manifest at a higher level in the API (e.g. `CqlSession<DropwizardMetrics>`, `CqlSession<MicrometerMetrics>`, etc). Ultimately I think switching metrics framework is a very rare edge case, so I'd rather keep the API simpler for the 99.9% regular ones.

Reply:
I think that in the general case, if you were plugging in a metrics framework, at some level that integration would provide access to the equivalent of a registry, so I think what you are doing here is fine.  Most users will be content with dropwizard metrics (I'd guess) and direct access to the registry here seems good.

Reply:
There might be a more generic solution.

This interface could stay agnostic of Dropwizard metrics:

1. Remove the `getRegistry()` method (it is not used by the driver anyway)
2. instead of `<T extends Metric>`, use simply `<T>`

Granted, people would have to cast to `DefaultMetrics` (maybe renamed to `DropwizardMetrics`?) to access the registry, but that looks cleaner to me: if someone wants to change for another metrics framework, it's just a matter of implementing another `MetricsFactory` (unless I'm missing something).

Reply:
> people would have to cast to DefaultMetrics

That's what I meant by "would have to manifest at a higher level in the API". Requiring a cast for a default feature is really bad IMHO. I'd rather not punish the 99.9%, and instead cause a minor inconvenience to the 0.1% that might want to change the metrics framework. Frankly the fact that this is even possible is a cherry on top of the cake already.

Reply:
I slightly disagree. A cast is still better imo, and I still like the idea that there is no "default" metric library in the driver, it just happens that we ship with one impl based on Dropwizard. But that's a detail and I don't want to spend more time on it, I'm fine if we leave things as they are.